CRM/Contribute/Tokens.php

```diff
@@ -18,70 +23,84 @@
  * At time of writing, we don't have any particularly special tokens -- we just
  * do some basic formatting based on the corresponding DB field.
  */
-class CRM_Contribute_Tokens extends \Civi\Token\AbstractTokenSubscriber {
+class CRM_Contribute_Tokens extends AbstractTokenSubscriber {
+
+  use CRM_Core_TokenTrait;
 
   /**
-   * Get a list of tokens whose name and title match the DB fields.
-   * @return array
+   * @return string
    */
-  protected function getPassthruTokens() {
-    return [
-      'contribution_page_id',
-      'receive_date',
-      'total_amount',
-      'fee_amount',
-      'net_amount',
-      'trxn_id',
-      'invoice_id',
-      'currency',
-      'contribution_cancel_date',
-      'receipt_date',
-      'thankyou_date',
-      'tax_amount',
-    ];
+  private function getEntityName(): string {
+    return 'contribution';
+  }
+
+  /**
+   * @return string
+   */
+  private function getEntityTableName(): string {
+    return 'civicrm_contribution';
+  }
+
+  /**
+   * @return string
+   */
+  private function getEntityContextSchema(): string {
+    return 'contributionId';
+  }
+
+  /**
+   * @var array
+   */
+  protected $paymentInstruments = [];
+
+  /**
+   * @var array
+   */
+  protected $financialTypes = [];
+
+  /**
+   * @var array
+   */
+  protected $contributionStatuses = [];
+
+  /**
+   * Get the basic tokens provided.
+   *
+   * @return array token name => token label
+   */
+  protected function getBasicTokens() {
+    if (!isset($this->basicTokens)) {
+      $this->basicTokens = CRM_Utils_Array::collect('title', CRM_Contribute_DAO_Contribution::fields());
+      $this->basicTokens['payment_instrument'] = ts('Payment Instrument');
+      $this->basicTokens['status'] = ts('Contribution Status');
+      $this->basicTokens['financial_type'] = ts('Financial Type');
+    }
+    return $this->basicTokens;
   }
 
   /**
-   * Get alias tokens.
+   * Returns a mapping of alias tokens to actual token.
+   * For example to support {activity.activity_id} in addition to {activity.id} we return ['activity_id => 'id]
    *
    * @return array
    */
   protected function getAliasTokens() {
     return [
       'id' => 'contribution_id',
```

Comment:
For activities, we have 'id' as a basic token and 'activity_id' as an alias.  For contributions, we have 'id' as an alias and 'contribution_id' as a basic token.  Can we standardise on one pattern and use that consistently?